File: mobile/src/main/java/player/playback/QueueManager.java
/*
 * Copyright (C) 2014 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package player.playback;

import android.content.res.Resources;
import android.graphics.Bitmap;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.media.MediaMetadataCompat;
import android.support.v4.media.session.MediaSessionCompat;

import com.example.android.uamp.R;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

import player.AlbumArtCache;
import player.model.MusicProvider;
import player.utils.LogHelper;
import player.utils.MediaIDHelper;
import player.utils.QueueHelper;

/**
 * Simple data provider for queues. Keeps track of a current queue and a current index in the
 * queue. Also provides methods to set the current queue based on common queries, relying on a
 * given MusicProvider to provide the actual media metadata.
 */
public class QueueManager {
    private static final String TAG = LogHelper.makeLogTag(QueueManager.class);

    private final MusicProvider mMusicProvider;
    private final MetadataUpdateListener mListener;
    private final Resources mResources;
    private final QueueHelper queueHelper;

    // "Now playing" queue:
    private List<MediaSessionCompat.QueueItem> mPlayingQueue;
    private int mCurrentIndex;

    public QueueManager(@NonNull MusicProvider musicProvider,
                        @NonNull Resources resources,
                        @NonNull MetadataUpdateListener listener,
                        @NonNull QueueHelper queueHelper
    ) {
        this.mMusicProvider = musicProvider;
        this.mListener = listener;
        this.mResources = resources;
        this.queueHelper = queueHelper;

        mPlayingQueue = Collections.synchronizedList(new ArrayList<MediaSessionCompat.QueueItem>());
        mCurrentIndex = 0;
    }

    public boolean isSameBrowsingCategory(@NonNull String mediaId) {
        String[] newBrowseHierarchy = MediaIDHelper.getHierarchy(mediaId);
        MediaSessionCompat.QueueItem current = getCurrentMusic();
        if (current == null) {
            return false;
        }
        String[] currentBrowseHierarchy = MediaIDHelper.getHierarchy(
                current.getDescription().getMediaId());

        return Arrays.equals(newBrowseHierarchy, currentBrowseHierarchy);
    }

    private void setCurrentQueueIndex(int index) {
        if (index >= 0 && index < mPlayingQueue.size()) {
            mCurrentIndex = index;
            mListener.onCurrentQueueIndexUpdated(mCurrentIndex);
        }
    }

    public boolean setCurrentQueueItem(long queueId) {
        // set the current index on queue from the queue Id:
        int index = queueHelper.getMusicIndexOnQueue(mPlayingQueue, queueId);
        setCurrentQueueIndex(index);
        return index >= 0;
    }

    public boolean setCurrentQueueItem(String mediaId) {
        // set the current index on queue from the music Id:
        int index = queueHelper.getMusicIndexOnQueue(mPlayingQueue, mediaId);
        setCurrentQueueIndex(index);
        return index >= 0;
    }

    public boolean skipQueuePosition(int amount) {
        int index = mCurrentIndex + amount;
        if (index < 0) {
            // skip backwards before the first song will keep you on the first song
            index = 0;
        } else {
            // skip forwards when in last song will cycle back to start of the queue
            index %= mPlayingQueue.size();
        }
        if (!queueHelper.isIndexPlayable(index, mPlayingQueue)) {
            LogHelper.e(TAG, "Cannot increment queue index by ", amount,
                    ". Current=", mCurrentIndex, " queue length=", mPlayingQueue.size());
            return false;
        }
        mCurrentIndex = index;
        return true;
    }

    public boolean setQueueFromSearch(String query, Bundle extras) {
        List<MediaSessionCompat.QueueItem> queue =
                queueHelper.getPlayingQueueFromSearch(query, extras, mMusicProvider);
        setCurrentQueue(mResources.getString(R.string.search_queue_title), queue);
        updateMetadata();
        return queue != null && !queue.isEmpty();
    }

    public void setRandomQueue() {
        setCurrentQueue(mResources.getString(R.string.random_queue_title),
                queueHelper.getRandomQueue(mMusicProvider));
        updateMetadata();
    }

    public void setQueueFromMusic(String mediaId) {
        LogHelper.d(TAG, "setQueueFromMusic", mediaId);
        String queueTitle = "";
        setCurrentQueue(queueTitle,
                queueHelper.getPlayingQueue(mediaId, mMusicProvider), mediaId);
        updateMetadata();
    }


    public void addToQueue(String uri) {
        List<MediaSessionCompat.QueueItem> newQueue = queueHelper.getPlayingQueue(uri, mMusicProvider);
        mPlayingQueue = newQueue;
        updateMetadata();
    }

    public MediaSessionCompat.QueueItem getCurrentMusic() {
        if (!queueHelper.isIndexPlayable(mCurrentIndex, mPlayingQueue)) {
            return null;
        }
        return mPlayingQueue.get(mCurrentIndex);
    }

    public int getCurrentQueueSize() {
        if (mPlayingQueue == null) {
            return 0;
        }
        return mPlayingQueue.size();
    }

    protected void setCurrentQueue(String title, List<MediaSessionCompat.QueueItem> newQueue) {
        setCurrentQueue(title, newQueue, null);
    }

    protected void setCurrentQueue(String title, List<MediaSessionCompat.QueueItem> newQueue,
                                   String initialMediaId) {
        mPlayingQueue = newQueue;
        int index = 0;
        if (initialMediaId != null) {
            index = queueHelper.getMusicIndexOnQueue(mPlayingQueue, initialMediaId);
        }
        mCurrentIndex = Math.max(index, 0);
        mListener.onQueueUpdated(title, newQueue);
    }

    public void updateMetadata() {
        MediaSessionCompat.QueueItem currentMusic = getCurrentMusic();
        if (currentMusic == null) {
            mListener.onMetadataRetrieveError();
            return;
        }
        final String musicId = currentMusic.getDescription().getMediaId();
        MediaMetadataCompat metadata = mMusicProvider.getMusic(musicId);
        if (metadata == null) {
            throw new IllegalArgumentException("Invalid musicId " + musicId);
        }

        mListener.onMetadataChanged(metadata);

        // Set the proper album artwork on the media session, so it can be shown in the
        // locked screen and in other places.
        if (metadata.getDescription().getIconBitmap() == null &&
                metadata.getDescription().getIconUri() != null) {
            String albumUri = metadata.getDescription().getIconUri().toString();
            AlbumArtCache.getInstance().fetch(albumUri, new AlbumArtCache.FetchListener() {
                @Override
                public void onFetched(String artUrl, Bitmap bitmap, Bitmap icon) {
                    mMusicProvider.updateMusicArt(musicId, bitmap, icon);

                    // If we are still playing the same music, notify the listeners:
                    MediaSessionCompat.QueueItem currentMusic = getCurrentMusic();
                    if (currentMusic == null) {
                        return;
                    }
                    String currentPlayingId = MediaIDHelper.extractMusicIDFromMediaID(
                            currentMusic.getDescription().getMediaId());
                    if (musicId.equals(currentPlayingId)) {
                        mListener.onMetadataChanged(mMusicProvider.getMusic(currentPlayingId));
                    }
                }
            });
        }
    }


    public interface MetadataUpdateListener {
        void onMetadataChanged(MediaMetadataCompat metadata);

        void onMetadataRetrieveError();

        void onCurrentQueueIndexUpdated(int queueIndex);

        void onQueueUpdated(String title, List<MediaSessionCompat.QueueItem> newQueue);
    }
}

File: mobile/src/main/java/player/playback/QueueManager.kt
/*
 * Copyright (C) 2014 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package player.playback

import android.content.res.Resources
import android.graphics.Bitmap
import android.os.Bundle
import android.support.v4.media.MediaBrowserCompat
import android.support.v4.media.MediaMetadataCompat
import android.support.v4.media.session.MediaSessionCompat
import com.example.android.uamp.R
import player.AlbumArtCache
import player.model.MusicProvider
import player.utils.MediaIDHelper
import player.utils.QueueHelper
import java.util.*

class QueueManager(private val musicProvider: MusicProvider,
                   private val mResources: Resources,
                   private val mListener: MetadataUpdateListener,
                   private val queueHelper: QueueHelper
) {
    private var mPlayingQueue: List<MediaSessionCompat.QueueItem>? = null
    private var mCurrentIndex: Int = 0
    internal val currentMusicMediaId
        get() = currentMusic?.description?.mediaId

    val currentMusic: MediaSessionCompat.QueueItem?
        get() = if (!queueHelper.isIndexPlayable(mCurrentIndex, mPlayingQueue)) {
            null
        } else mPlayingQueue?.get(mCurrentIndex)

    val currentQueueSize: Int
        get() = mPlayingQueue?.size ?: 0

    val playlist: List<MediaBrowserCompat.MediaItem>
        get() = queueHelper.playlist

    init {
        mPlayingQueue = Collections.synchronizedList(ArrayList())
        mCurrentIndex = 0
    }

    private fun setCurrentQueueIndex(index: Int) {
        val size = mPlayingQueue?.size ?: 0
        if (index >= 0 && index < size) {
            mCurrentIndex = index
            mListener.onCurrentQueueIndexUpdated(mCurrentIndex)
        }
    }

    fun setCurrentQueueItem(queueId: Long): Boolean {
        val index = queueHelper.getMusicIndexOnQueue(mPlayingQueue, queueId)
        setCurrentQueueIndex(index)
        return index >= 0
    }

    fun skipQueuePosition(amount: Int): Boolean {
        var index = mCurrentIndex + amount
        if (index < 0) {
            index = 0
        } else {
            index %= mPlayingQueue?.size ?: 0
        }
        if (!queueHelper.isIndexPlayable(index, mPlayingQueue)) {
            return false
        }
        mCurrentIndex = index
        return true
    }

    fun setQueueFromSearch(query: String, extras: Bundle): Boolean {
        val queue = queueHelper.getPlayingQueueFromSearch(query, extras, musicProvider)
        setCurrentQueue(mResources.getString(R.string.search_queue_title), queue)
        updateMetadata()
        return queue != null && !queue.isEmpty()
    }

    fun setRandomQueue() {
        setCurrentQueue(mResources.getString(R.string.random_queue_title),
                queueHelper.getRandomQueue(musicProvider))
        updateMetadata()
    }

    fun setQueueFromMusic(mediaId: String) {
        val queueTitle = ""
        setCurrentQueue(queueTitle, queueHelper.getPlayingQueue(mediaId, musicProvider), mediaId)
        updateMetadata()
    }


    fun addToQueue(uri: String) {
        val newQueue = queueHelper.getPlayingQueue(uri, musicProvider)
        mPlayingQueue = newQueue
        updateMetadata()
    }

    @JvmOverloads protected fun setCurrentQueue(title: String, newQueue: List<MediaSessionCompat.QueueItem>?,
                                                initialMediaId: String? = null) {
        mPlayingQueue = newQueue
        var index = 0
        if (initialMediaId != null) {
            index = queueHelper.getMusicIndexOnQueue(mPlayingQueue, initialMediaId)
        }
        mCurrentIndex = Math.max(index, 0)
        mListener.onQueueUpdated(title, newQueue)
    }

    fun updateMetadata() {
        val currentMusic = currentMusic
        if (currentMusic == null) {
            mListener.onMetadataRetrieveError()
            return
        }
        musicProvider.getMusic(currentMusic.description?.mediaId)?.let {
            updateMedtadataInternal(it, currentMusic)
        }
    }

    private fun updateMedtadataInternal(mediaMetadataCompat: MediaMetadataCompat, currentMusic: MediaSessionCompat.QueueItem) {
        mListener.onMetadataChanged(mediaMetadataCompat)
        if (mediaMetadataCompat.description.iconBitmap == null && mediaMetadataCompat.description?.iconUri != null) {
            updateArt(mediaMetadataCompat, currentMusic)
        }
    }

    private fun updateArt(mediaMetadataCompat: MediaMetadataCompat, currentMusic: MediaSessionCompat.QueueItem) {
        AlbumArtCache.getInstance().fetch(mediaMetadataCompat.description?.iconUri?.toString(), object : AlbumArtCache.FetchListener() {
            override fun onFetched(artUrl: String, bitmap: Bitmap, icon: Bitmap) {
                musicProvider.updateMusicArt(mediaMetadataCompat.description?.mediaId, bitmap, icon)
                val mediaId = currentMusic?.description?.mediaId ?: return
                val currentPlayingId = MediaIDHelper.extractMusicIDFromMediaID(mediaId)
                if (mediaId == currentPlayingId) {
                    musicProvider.getMusic(currentPlayingId)?.let { mListener.onMetadataChanged(it) }
                }
            }
        })
    }

    interface MetadataUpdateListener {
        fun onMetadataChanged(metadata: MediaMetadataCompat)

        fun onMetadataRetrieveError()

        fun onCurrentQueueIndexUpdated(queueIndex: Int)

        fun onQueueUpdated(title: String, newQueue: List<MediaSessionCompat.QueueItem>?)
    }
}
